### ISSUE TITLE
Set MaterialAboutToggleItem state programmatically

### ISSUE THREAD

**Issue by Gaket** [daniel-stoneuk/material-about-library#77]
Hi! Is it possible to set a toggle state programmatically?

It looks like `MaterialAboutToggleItem` doesn't have such method.

I have a case when I should set toggle back depending on user answer. Like if user toggles “set fullscreen”, then we show him permission dialog and if the user denies to give the permission, we uncheck the toggle back programmatically.

**Comment by daniel-stoneuk** [comment]
MaterialAboutToggleItem? Don't think I've created one of these - is it in a forked library. I've actually wanted to implement some items like this so that people can use the library as a settings page also.

**Comment by daniel-stoneuk** [comment]
Hi, this functionality has been implemented in a fork at by @filol here: https://github.com/filol/material-preference-library

**Comment by Robyer** [comment]
@daniel-stoneuk Perhaps you can close this issue as it was merged?

**Comment by daniel-stoneuk** [comment]
@Robyer we haven't merged it into the main branch yet since it's performance is pretty terrible.

**Comment by Robyer** [comment]
@daniel-stoneuk Aha, I saw on the MaterialPreferenceLibrary notice that that library is deprecated now as it was merged to MAL.
And can you be more specific? What kind of performance issues are there?

**Comment by Robyer** [comment]
Also, I'm comparing the changes between your branch and master and I noticed some weird things:

- why is there new INTERNET permission in AndroidManifest?

- why is there `areContentsTheSame` method that calls getDetailString() to create big string representations of the objects and then compares the two strings? Why not just properly implement equals() method in first place?

- if the performance issues you are talking about are when you set checked listener and then deny the checked state as in that example, it is caused by this: we are setting `this.aCheckBox.setOnCheckedChangeListener(`, but this is called always, when the internal checked state changed, i.e. even when it is programatically changed by calling "setChecked". So in example [here](https://github.com/daniel-stoneuk/material-about-library/compare/preferences#diff-9799e528da99265803f7a9c4ba19f383R108) we are inside checked change listener, and here we call `setChecked(true)` again, which will as a result call our checked change listener again, where we will again call it... Also notice that we return false from this method (to notify that we don't want to allow the change). So we may just remove that manual call to `switchItem.setChecked(true);` and theoretically it should work. BUT it is implemented [here](https://github.com/daniel-stoneuk/material-about-library/compare/preferences#diff-6262c2551a1b291439441e24b59f88afR315) with the same infinite recursion problem - if user returns false, then we call the `setChecked(!isChecked)` which in turn again calls our own listener, where we call that again... (second infinite recursion). Thing why it won't end up in app crash is that there is the check for `buttonView.isPressed()`, so the infinite recursion is going on as long as you hold your finger on the screen. So that code must be reworked...

**Comment by daniel-stoneuk** [comment]
> why is there areContentsTheSame method that calls getDetailString() to create big string representations of the objects and then compares the two strings? Why not just properly implement equals() method in first place? 

Yep spot on with the edit, and you're right. We should implement a function to compare just the visuals - generating the strings and then comparing them is probably quite slow and wasteful. I've created issue #91. Probably won't get round to fixing this until after my University interview in a couple weeks.

The performance issues I was noticing was during scrolling. Lots of frames are dropped, which doesn't really give a good user experience. Because of this, I didn't clean up the PR too much & kept it in a separate branch. I think part of this is due to the inherent performance issues that come with using nested RecyclerViews - I knew from the start that this wouldn't be a great idea for scalability, however it seemed most practical & I didn't want to include library like [groupie](https://github.com/lisawray/groupie), which might struggle to display a cardview beneath the items. Another perk of the nested RVs is the ability to use custom adapters like LicenseAdapter.

Would you say the fixes you've made are ready to be merged? Thanks very much for the commits - as I mentioned, I don't have much free time right now but in a couple weeks should be able to put more work into it if required.

**Comment by Robyer** [comment]
@daniel-stoneuk As I now tried it, the bad performance during scrolling really seems to be caused by the nested RecyclerViews - it can be seen when you scroll down slowly, when the Custom Adapter should be drawn, it lags for some hundreds of milliseconds. So problem is not in the checkbox/toggle.

I think the preferences branch could be merged (as it is not the cause of bad scrolling performance), but the feature should be improved in the future - e.g. configure on click listener to whole row so it will change checkbox/toggle when clicking anywhere on the row. Or also I think changing the row checked state programatically won't work. But that is related only to new 2 items, everything in library should work same as before, that why I think you can merge it.
